Layer Based Rendering of Gadgets

mattigruener opened the pull request:
Hey John,

Here is the first attempt regarding rejiggering the Gadget rendering order. I think it pretty much follows what we discussed? A few things to note:

- The `GraphGadget` implementation isn't as clean as we'd like. `Backdrops` mess a little with our plan, don't they? I've updated the comment in the code accordingly, but maybe I'm missing a simple solution.
- I've updated the rendering bits for `PlugAdders` and `StandardNodules` and I think they behave much better now? :) I checked for other spots in the code where we are looking for the highlighting state of `Gadgets` because I'm assuming that whenever something is highlighted it should get rendered last, but couldn't find anything else. Probably worth keeping in mind for the future.
- There's potential to be a bit more DRY, especially in `Gadget`'s `render*` methods. Not sure if it's worth pulling out some of the duplicated code.
- I wasn't sure about C++11 features, I'm using both auto and range-based for in GraphGadget's rendering code. That's probably ok?

For anyone following along - this PR is meant to be the foundation for some of the additional work we're doing in the never-ending https://github.com/GafferHQ/gaffer/pull/2201. In that PR we'll update the label rendering and some of work done here will help with that.

Thanks :)

Matti.

johnhaddon commented:
> The GraphGadget implementation isn't as clean as we'd like. Backdrops mess a little with our plan, don't they? I've updated the comment in the code accordingly, but maybe I'm missing a simple solution.

Gah! I guess a more general solution might be something more along these lines?

```
Gadget : ...
{
       
       virtual int minLayer(); // Returns 0 by default. -1 would mean there's an underlay
       virtual int maxLayer(); // Returns 0 by default. 1 would mean there's an overlay
       virtual void renderLayer( int layer, const Style *currentStyle );

}
```

Lot's of details to work out there of course, particularly with the necessary implementations for taking children into account, and not introducing too much overhead. Let's chat about it...

> There's potential to be a bit more DRY, especially in Gadget's render* methods. Not sure if it's worth pulling out some of the duplicated code.

My guess would be that the `renderLayer()` approach above might help with that, because there's only a single `renderLayer()` method?

> I wasn't sure about C++11 features, I'm using both auto and range-based for in GraphGadget's rendering code. That's probably ok?

Yeah, I think ranged-based for is clearly more readable in these situations. You do need to be a bit careful with auto though :

`for( auto child : children() )`

As I understand it, this is copy constructing an element of `children()` into `child` each time through the loop. And `children()` contains `GraphComponentPtr`, and copying `intrusive_ptr` has overhead (atomic reference count updates). We already know we own the children (they're our children after all), so have no need to manipulate reference counts - so I think either `const GraphComponentPtr &` or `const auto &` are what you want here.

More generally, I'm trying to figure out when and where I'm comfortable using `auto`. Maybe I'm just a fuddy duddy, but I find the "almost always auto" guideline a bit hard to swallow. Because I don't use a fancy IDE, I kindof want the type to be obvious from the code itself. I like `for( auto i = begin( c )...`
because it's obvious that it'll be an iterator, and it makes things more readable. And I might even get to like `auto x = runTimeCast<Gadget>( child.get() )` because the type is in the right hand side of the expression anyway. But I think I'll be dipping my toes at first rather than diving right in...

mattigruener commented:
Hey John,

I just pushed the code for the new attempt.

So, I deviated a little from your suggestion that included the `minLayer()` and `maxLayer()` methods and only implemented the `renderLayer()` part of it as a first, stupid solution. I played with the min/max idea a little, but felt that it doesn't really add anything and only increases API surface area and adds work for Gadget implementers that doesn't really pay off? My understanding was that we'd go through all Gadgets in the GraphGadget to find the range of layers we need to render using the min/max methods and then call renderLayer for all of them. I think we get to a point where we render something into every layer really quickly and at that point the min/max gathering is merely overhead, isn't it?

I personally also felt that having a dynamic number of layers is not really needed? To me it seems more explicit and easier to understand if we have a fixed number (in the current implementation that would be in the RenderLayer enum) and have the Gadgets pick one or multiple of these layers to render into. It seems like that would make it a little easier for Gadget developers? I'd find it a little odd if each Gadget would need to understand the grand scheme of how things are rendered and then decide for itself where to put itself in terms of layering - my intuition would be to name (or just add a comment to) the layers to make them a little more descriptive, "ConnectionLayer" for example, so that all Gadgets that want to render a Connection (PlugAdders, ConnectionGadgets, Nodules...) just render into it. Does that make sense? I understand that the layers in the current implementation are not just for the GraphGadget and its children, but having a central location that decides how this stuff is composited seems nicer to me than having every Gadget decide on its own? Maybe that's just me coming to an existing code base, having a first place to look at would help newbees? :)

Also, the open question with the current implementation is how to handle the Python binding. I think I will have to update a couple of tests if we do something like what I'm proposing here.

Alright, some more stuff to talk about, I guess. I think performance is fine btw, even with the overhead of a bunch of unnecessary function calls. I created a synthetic scene with lotsa nodes and it doesn't seem noticeably slower than rendering the same scene with the previous implementation.

Thanks! :)

Matti.

mattigruener commented:
Papertrail: This PR, in conjunction with PR https://github.com/GafferHQ/gaffer/pull/2201, will take care of issue https://github.com/GafferHQ/gaffer/issues/674, as well.

johnhaddon commented:
> I just pushed the code for the new attempt.

Cool - thanks!

> I played with the min/max idea a little, but felt that it doesn't really add anything and only increases API surface area and adds work for Gadget implementers that doesn't really pay off?

Yeah, I agree it was pretty clunky - I was trying to generalise too far. Thanks for pushing for the simpler solution.

> I personally also felt that having a dynamic number of layers is not really needed? 

The only thing I know of that could take advantage of it is drawing nested backdrops. Currently the draw order works only if the inner backdrop was created after the outer backdrop, otherwise it draws underneath. I'm pretty sure we have a ticket open for it somewhere, which is what made me think of the minLayer/maxLayer thing.

Even with arbitrary numbers of layers though, we'd still need backdrops to be aware of each other to come up with a suitable draw order, and the draw order might need to change as they're dragged around. So perhaps it's easier to use a single layer for backdrops, and then have special case drawing code whereby nested backdrops defer their drawing to the outer backdrop. A little bit of special case trickery in that one instance is better than complicating the API _everywhere_ I think.

> my intuition would be to name (or just add a comment to) the layers to make them a little more descriptive, "ConnectionLayer" for example, so that all Gadgets that want to render a Connection (PlugAdders, ConnectionGadgets, Nodules...) just render into it. Does that make sense? I understand that the layers in the current implementation are not just for the GraphGadget and its children...

I think it'd be a mistake to name the layers purely based on usage in the GraphGadget. Imagine implementing a pixel inspector for the viewer but having to render into the NoduleLayer - weird right? Perhaps we could do this though?

```
class Gadget ...
{

  enum class Layer
  {
     Back = -2,
     MidBack = -1,
     ... 
  };

};

class GraphGadget ...
{
  
  enum class Layer
  {
     Backdrops = Gadget::Layer::Back,
     Connections = Gadget::Layer::MidBack,
     ...
  }

};

```

I don't know if we get into problems casting one to the other that way, but if we can make it work it seems like the best of both worlds to me.

> I think performance is fine btw, even with the overhead of a bunch of unnecessary function calls. I created a synthetic scene with lotsa nodes and it doesn't seem noticeably slower than rendering the same scene with the previous implementation.

Great - that was my only real worry with this layering approach.

Taking your "minimising surface area" thing further, I almost wonder if we need a public `renderLayer()` method at all. I'm also a bit confused about other elements of the current implementation - why does GraphGadget have explicit code to draw all it's children still for instance? Shouldn't that be handled by the Gadget base class?

mattigruener commented:
Hey John,

some new stuff to look at. Really just cleaning up some bits. Do you find this less confusing now? :)

Re: nested Backdrops

Would it make sense to have nested Backdrops be handled by adding them as child to Backdrops they're dragged onto? And then have Gadget::doRenderLayer() render them after having rendered the parent? Not sure if that would mess up anything else if they're no first class citizens in the GraphGadget anymore but only exist through parent/child hierarchies...

> I think it'd be a mistake to name the layers purely based on usage in the GraphGadget. 

Yes, for sure. How do you like the namespace based solution I've used here? I tried the enum-only approach and the casting felt pretty clumsy...

Probably getting a little closer with this one...

Thanks! :)

Matti.

johnhaddon commented:
Hey Matti,
I'm having a bit of trouble with the graph below. The top three nodes behave as expected, but the ShaderAssignment plugs are elusive - as soon as I hover over them they kindof flicker and disappear...
Cheers..
John

```
import Gaffer
import GafferScene
import IECore

Gaffer.Metadata.registerNodeValue( parent, "serialiser:milestoneVersion", 0, persistent=False )
Gaffer.Metadata.registerNodeValue( parent, "serialiser:majorVersion", 41, persistent=False )
Gaffer.Metadata.registerNodeValue( parent, "serialiser:minorVersion", 0, persistent=False )
Gaffer.Metadata.registerNodeValue( parent, "serialiser:patchVersion", 0, persistent=False )

__children = {}

__children["Plane"] = GafferScene.Plane( "Plane" )
parent.addChild( __children["Plane"] )
__children["Plane"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = IECore.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Plane"]["__uiPosition"].setValue( IECore.V2f( -10.6334944, 3.3756671 ) )
__children["Sphere"] = GafferScene.Sphere( "Sphere" )
parent.addChild( __children["Sphere"] )
__children["Sphere"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = IECore.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Sphere"]["__uiPosition"].setValue( IECore.V2f( 2.36650562, 3.37566662 ) )
__children["Group"] = GafferScene.Group( "Group" )
parent.addChild( __children["Group"] )
__children["Group"]["in"].addChild( GafferScene.ScenePlug( "in1", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"]["in"].addChild( GafferScene.ScenePlug( "in2", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = IECore.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"]["__uiPosition"].setValue( IECore.V2f( -2.63333082, -4.7892642 ) )
__children["ShaderAssignment"] = GafferScene.ShaderAssignment( "ShaderAssignment" )
parent.addChild( __children["ShaderAssignment"] )
__children["ShaderAssignment"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = IECore.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ShaderAssignment"]["__uiPosition"].setValue( IECore.V2f( -2.63300228, -12.9550648 ) )
__children["Group"]["in"]["in0"].setInput( __children["Plane"]["out"] )
__children["Group"]["in"]["in1"].setInput( __children["Sphere"]["out"] )
__children["ShaderAssignment"]["in"].setInput( __children["Group"]["out"] )

del __children
```

mattigruener commented:
Oh cool, I'm having a look as well and the reason seems to be that the doRenderLayer() method for that StandardNodeGadget is only called for the main/"nodes" layer. That's why highlighted nodules just don't get rendered and the whole thing disappears. I'll have a look at your changes once you got them online. Thanks :)

mattigruener commented:
You probably fixed the python binding, hey? I think it's this line in GadgetBinding.h that is causing it:

`WrappedType::doRenderLayer( GafferUI::RenderLayer::Main, style );`

Seems all the wrapped nodes only render stuff to that one layer at the moment. I think that was a relict of something I was doing before, sorry.

johnhaddon commented:
> You probably fixed the python binding, hey?

I think so, yep - got some changes there at least.

johnhaddon commented:
Hey Matti, here's my branch :

https://github.com/johnhaddon/gaffer/tree/mattigruener-overUnderDrawing

I've built on top of yours, changed a few names to be less verbose, simplified a few things, fixed a few bugs and made another simplification that your changes made possible. If this all looks OK to you, could you update your branch with my changes, squashing each of my commits into the corresponding original commit of yours? Then hopefully we're in a good shape to merge.

Thanks for everything on this. The "fixed number of layers" design you proposed feels like a great balance between flexibility and simplicity, and it's nice to see we've already been able to fix a few bugs. I've got no doubt that it's also going to simplify the implementation of other UI features in the future...

Cheers...
John

mattigruener commented on src/GafferUI/ViewportGadget.cpp:
```diff
@@ -250,7 +250,11 @@ void ViewportGadget::gadgetsAt( const Imath::V2f &rasterPosition, std::vector<Ga
 		s->bind();
-		Gadget::doRender( s );
+
+		for( int layer = (int)Layer::Back; layer < (int)Layer::Last; ++layer )
```
So you don't like auto here, I guess? If we want to do the loop like this, should we fix it in that other spot where it comes up too and then potentially also remove that operator override we're currently dragging along to allow `++layer` without any casts? I'll mark the other bit in the code where we do the same...

johnhaddon replied:
Sorry, I thought I'd put this back to the way you had it originally. I was experimenting to see what it felt like to remove the `operator++` (no problem with auto btw). I kindof liked that removing it meant it was impossible to generate invalid values without a cast, and also that the casts were now isolated in the loop where it was easy to verify that they stayed within range. So, we should either remove `operator++` and update Gadget, or replace this with the same code as in Gadget. What do you think?

mattigruener commented on src/GafferUI/Gadget.cpp:
```diff
@@ -254,7 +254,15 @@ Imath::M44f Gadget::fullTransform( const Gadget *ancestor ) const
-void Gadget::render( const Style *currentStyle ) const
+void Gadget::render() const
+{
+	for( auto layer = Layer::Back; layer < Layer::Last; ++layer )
```
See other comment, how to handle the loop here?

mattigruener commented on include/GafferUI/Gadget.h:
```diff
@@ -341,6 +354,8 @@ class Gadget : public Gaffer::GraphComponent
 typedef Gaffer::FilteredChildIterator<Gaffer::TypePredicate<Gadget> > GadgetIterator;
 typedef Gaffer::FilteredRecursiveChildIterator<Gaffer::TypePredicate<Gadget> > RecursiveGadgetIterator;
 
+inline Gadget::Layer &operator++( Gadget::Layer &x ) { return x = (Gadget::Layer)(((int)(x) + 1)); };
```
Potentially superfluous if we don't do the `auto layer; ++layer` gubbins anymore?

johnhaddon commented:
> I've integrated all your changes by naively squashing them into the respective previous commits.

Thanks. Could you add back the "Fixes #674" comment into StandardNodule? That'll close the ticket for us automatically when we merge, and link the issue to the commit. Then once you've decided which way to go with the `operator++` I think we're good to go. Let me know if you want me to tidy up my own mess there...

johnhaddon commented:
> Would it make sense to have nested Backdrops be handled by adding them as child to Backdrops they're dragged onto? And then have Gadget::doRenderLayer() render them after having rendered the parent? Not sure if that would mess up anything else if they're no first class citizens in the GraphGadget anymore but only exist through parent/child hierarchies...

I think whatever we do for the nested backdrops will be a bit funky, because there's nothing explicit in their representation that says they're nested, so we need to infer an order from their spatial relationships (unlike boxes/references which specify a natural hierarchical nesting). I think it would be a bit odd (not to say tricky) to try to transform the spatial relationship into a parenting one in the GraphGadget, especially since we've just simplified GraphGadget so that it's independent of the layering for everything else. I'm inclined to keep the backdrop nesting as a dirty secret that is handled entirely in the backdrop drawing code - it's only a visual tweak after all, so the less knock on effects there are in the rest of the code, the better.

mattigruener commented:
Hey John :)

I fixed the message for that one commit and pushed changes for the for loops. I've kept the `operator++` because I personally feel like the casting is a little messy and confusing - might just be me. Having the operator messing up just one spot in the code (Gadget.h) and not every spot where we'll have to loop through the layers seems preferable to me... I also removed `auto` and use the type explicitly. Using `auto` doesn't save much typing and `Layer` is more explicit... I'm happy to change things back if you feel more strongly about it - I don't really.

Great to get this one wrapped up :)

Matti.

johnhaddon commented:
Thanks Matti - merged. My thoughts on the `operator++` were that the looping is only needed inside Gadget.cpp (I think we can actually remove it from ViewportGadget.cpp), so following the principle of least privilege it'd make sense not to provide an operator which could be used to generate invalid layers elsewhere. I think we could even make Layer protected if it weren't for the fact that we wanted to make some aliases to it in GraphGadget.h. Anyway, this is trivia...

mattigruener commented:
Ok cool, I'll keep it in mind and if there's an opportunity to clean it up more, I'll do it. Thanks for your help with this. One PR down! :)

johnhaddon commented:
> Thanks for your help with this.

My pleasure. Thanks for your patience :)
